### ISSUE TITLE
Morphing text to fit inside rect

### ISSUE THREAD

**Issue by mikkelee** [pymupdf/PyMuPDF#259]
Final question for the day :)

I have a word and a rectangle, and I'm trying to use `insertTextbox()` to put it on the page. However, the word is too wide for the box. I thought I could fix this by using the `morph` parameter, but I must be doing something wrong.

```python
word = 'Efterlysninger.'
rect = fitz.Rect(97.25, 236.19651794433594, 161.28929138183594, 248.8695068359375)

textwidth = fitz.getTextlength(word)
pivot = fitz.Point(rect.x0, rect.y0 + rect.height/2) #middle of left side
matrix = fitz.Matrix(textwidth/rect.width, 1.0)
page.drawRect(rect, color=(0,0,1))
page.drawRect(rect, color=(1,0,1), morph=(pivot, matrix))
r1 = page.insertTextbox(token.rect, f'{word}', color=(0,1,0))
r2 = page.insertTextbox(token.rect, f'{word}', color=(1,0,0), morph=(pivot, matrix))
print(f'{word} - {textwidth} {rect.width} - morph: {pivot}, {matrix} - {r1} {r2}')
```

Result:
`Efterlysninger. -- 70.30099487304688 64.03929138183594 -- morph: Point(97.25, 242.53301239013672), Matrix(1.097779087745918, 0.0, 0.0, 1.0, 0.0, 0.0) -- -24.727011108398415 -24.727011108398415`

![skaermbillede 2019-02-28 kl 17 47 02](https://user-images.githubusercontent.com/644671/53583010-f42db680-3b80-11e9-8e8c-1e2607d0775b.png)

As far as I can tell, my rect was indeed made wider but it was still too small for the text by the same amount... I assume the text was morphed at the same time, but then what is the solution?

**Comment by JorjMcKie** [comment]
The method returns a negative float if the rect **height** is insufficient. The returned number should in case be the amount of missing height (if negative) or surplus height.
If a single word is longer than the rect **width**, the method cuts the word in pieces, which are then put on new lines until the word is exhausted. So, a single word may need more one line.
You could try and make the font smaller also or just make the rect really large (height) and look at the returned float to later eventually reduce again.
The rect serves no other purpose than the explained one - it does not enter the PDF itself. So you can define any number of new rects for new text boxes, each subsequent one having a top coordinate where a previous text ended, etc.

The (new) method ``getTextlength`` should be ok to take for the required rect width (if it is only that word to fit in. But rect height must still be sufficient: I am using fontsize * 1.2 for the height of 1 line.
If you do not specify fontsize in getTextlength, some default is take (I believe 11). To make sure, I would specify it - consistently with the insertTextbox value.

**Comment by JorjMcKie** [comment]
Ah - the morph parameter plays no role in all of this. You can just change the final appearance in wild ways (tilting, rotating, mirroring ...).
How this works can be seen [e.g. here](https://github.com/JorjMcKie/PyMuPDF-Utilities/blob/master/animations/morph-demo1.py) - you need ``PySimpleGUI`` to display the endless loops in there.

**Comment by mikkelee** [comment]
Hmm I must be doing something wrong still.

As shown, the word Sophus is too tall, but the others are not. They're all inserted (and width calculated) with a `fontsize=rect.height*0.7`

```
Sophus
 -- rect: Rect(110.75, 263.146484375, 140.69210815429688, 275.8194885253906)
 -- calc.width: 30.081911087036133 rect.width: 29.942108154296875
 -- rect.height: 12.673004150390625 result: -17.48874572753914
Lauritz
 -- rect: Rect(145.35000610351562, 263.146484375, 178.8108367919922, 275.8194885253906)
 -- calc.width: 26.622180938720703 rect.width: 33.46083068847656
 -- rect.height: 12.673004150390625 result: 3.8019012451171648
Viggo
 -- rect: Rect(186.14999389648438, 263.146484375, 212.08653259277344, 275.8194885253906)
 -- calc.width: 22.68341064453125 rect.width: 25.936538696289062
 -- rect.height: 12.673004150390625 result: 3.8019012451171648
Nepken,
 -- rect: Rect(217.35000610351562, 263.146484375, 253.52764892578125, 275.8194885253906)
 -- calc.width: 33.035987854003906 rect.width: 36.177642822265625
 -- rect.height: 12.673004150390625 result: 3.8019012451171648
```

But the rects have identical heights and I can't see an issue with ascenders/descenders either, all words have them.

In fact, it looks like every single word is either too tall by ~17.5 or short enough by ~3.8

**Comment by mikkelee** [comment]
Figured it out!

Turns out the box being slightly too narrow for _Sophus_ caused it to be broken up as `\nSophu\ns` which is 2 text-lines too tall for the box, in other words 2*(rect.height*0.7) = 2*8.87 = ~17.5. Making the width of the rect `max(textwidth, rect.width)` solved it.

**Comment by mikkelee** [comment]
Augh, I was a bit too fast on the trigger. Even when using fitz.getTextlength as the width of a rect, there are some cases where the word is too wide when inserted (using the same font+size).

Examples:
meget
betale.
Ds.
saaledes:
klippet,
Johan
Søren
Skomager

I've fixed it by adding 5.0 to the width, but could there be a bug in how the width is calculated?

**Comment by mikkelee** [comment]
This comment was really just to say that I have my use case fully working! Prioritize your time/effort as you please :)

One thing strikes me: `Sophus` was turned into `\nSophu\ns` which is probably sub-optimal --- I don't think a leading linefeed is ever the desired outcome when inserting text in a box that is too narrow (it will hide the text if the box isn't tall enough for the following lines). The second linefeed makes sense and I have no objection there.

The problem was identical in the examples from my previous comment (leading linefeed, beginning of word, linefeed, rest of word).

**Comment by JorjMcKie** [comment]
Congrats that you at least have an acceptable status quo.
I will look into this ``Sophus`` example over the weekend.
Have to apologize for my high latency ... but my internet is simply horrible in these hard times for Venezuela!

For a technical background: the calculation of ``getTextlength`` is done by the MuPDF base library itself (as a monolithic C function).
The calculations required during text fitting in boxes are my own humble making (written in pure Python that piece). Although this is also is based on MyPDF logic in the end (on a by character level), there is at least a theoretical chance for deviations.
So I have best personal interest to learn what is going on in your case.

**Comment by JorjMcKie** [comment]
This is the explanation:
fitz.getTextlength("Sophus") is 37.30099868774414 (on Win 10)
If choosen a rectangle of exactly this width, the word will not fit - there just has to some allowance. In this case choosen a width of 38 was sufficient:

Rect width has exact word length:
![grafik](https://user-images.githubusercontent.com/8290722/53668333-fa3b9a00-3c49-11e9-9e09-e777cbc8de17.png)

Rect has width of 38:
![grafik](https://user-images.githubusercontent.com/8290722/53668394-30791980-3c4a-11e9-8560-c9a6652fae4b.png)

So this explains the final "s" being put on a neew line. But the logic of the method deliberately is:
Whenever a word is longer than the box width, start a new line for that word and start filling available room with the word's characters.

So everything works as designed here.
Maybe not completely / well enough documented ...

**Comment by JorjMcKie** [comment]
Although ... I could spare me a leading ``\n`` if my position is already the leftmost

**Comment by mikkelee** [comment]
Aha, that makes sense. Thank you for the explanation of the details :)

Best wishes to you and yours, and have a good weekend!

**Comment by JorjMcKie** [comment]
Same to you!
And thanks for your patience.

**Comment by JorjMcKie** [comment]
I made the announced change in insertTextbox (no line break for first line of a text box).

I also investigated further, what actually caused the different judgement of word length in ``getTextlength`` and ``insertTextbox``.
And the answer seems to be: **rounding issues!**
If I calculate the length of "Sophus" in 2 different ways, I get the following differences:

```python
>>> l1 = fitz.getTextlength("Sophus")
>>> l2=0.0
>>> for c in "Sophus":
	l2 += fitz.getTextlength(c)

>>> l1, l2
(37.30099868774414, 37.30100059509277)
>>>
```
There is a tiny difference between `l1 `and `l2`.
If I use `l2 `as the rectangle width instead of `l1`, the **word will fit!** The red rectangle has width `l2`, the blue one 38:

![grafik](https://user-images.githubusercontent.com/8290722/53680332-5b965400-3cb0-11e9-8c75-30d424281622.png)

In wrapping up: with your help, the findings are:
1. `insertTextbox `works correctly, but needs to suppress leading line breaks.
2. when using `getTextlength `results for calculating text box widths, either add some cushion (e.g. add the length of 1 space character, i.e. use length of `"Sophus "`), or (worse performance way) use the sum of lengths of single characters.

I will update my documentation accordingly.
Thanks again!!
